### PR TITLE
[feature](bangc-ops): improve box_iou_rotated.

### PR DESCRIPTION
**Pull request by DanieeelLiu**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

Improve the performance on 500.

## 2. Modification

modified:   bangc-ops/kernels/box_iou_rotated/box_iou_rotated_aligned.h
modified:   bangc-ops/kernels/box_iou_rotated/box_iou_rotated_nonaligned.h
modified:   bangc-ops/kernels/box_iou_rotated/box_iou_rotated_utils.h
modified:   bangc-ops/kernels/box_iou_rotated/box_iou_rotated.cpp
modified:   docs/bangc-docs/design_docs/box_iou_rotated/box_iou_rotated.md
new file:   docs/bangc-docs/design_docs/box_iou_rotated/res/cosine1.png
new file:   docs/bangc-docs/design_docs/box_iou_rotated/res/sort_vertices1.png

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff1
    - [ ] float32 mlu diff1 <= 1e-5
    - [x] float32 mlu diff1 <= 3e-3
    - [ ] float16 mlu diff1 <= 3e-3
  - diff2
    - [ ] float32 mlu diff2 <= 1e-5
    - [x] float32 mlu diff2 <= 3e-3
    - [ ] float16 mlu diff2 <= 3e-3
  - diff3
    - [ ] mlu diff3 == 0
    - [ ] mlu diff3_1 == 0
    - [ ] mlu diff3_2 == 0
- dynamic threshold
  - [ ] diff1: mlu diff1 <= max(baseline diff1 * 10, static threshold)
  - [ ] diff2: mlu diff2 <= max(baseline diff2 * 10, static threshold)
  - [ ] diff3: mlu diff3 <= max(baseline diff3 * 10, static threshold)
    - float32, threshold = 1e-5
    - float16, threshold = 1e-3

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [x] MLU370
  - [x] MLU590
- Job types
  - [ ] BLOCK
  - [x] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test
If you have checked the following items, please tick the relevant box.

- [ ] Data type test (e.g. float32/int8)
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [ ] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [x] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

all release_test/release_temp/benchmark passed on 300 & 500.

#### 3.2.2 Parameter Check

Test Point-1: `MLU300 box num should less than 10,000,000`. Acceptance Standard: `Normal error`.
```bash
[2023-10-18 15:0:47] [MLUOP] [Warning]:Check failed: mt->shape_count == shape_count. "Parser: found shape count is not equal to value " "size.(mt->shape_count is value_size)"
[2023-10-18 15:0:48] [MLUOP] [Error]:[mluOpBoxIouRotated] Check failed: If it is aligned mode, on MLU300 box1_desc->dims[0] should less than or equal to 10,000,000 . But now is 429496730.
[2023-10-18 15:0:48] [MLUOP] [Error]:"MLUOP_STATUS_BAD_PARAM in mluOpBoxIouRotated(handle_, mode, aligned, box1, dev_box1, box2, dev_box2, ious, dev_ious)"
[2023-10-18 15:0:48] [MLUOP] [Warning]:Executor: reset cnrt and go on running, this may caused by cnrt failed.
/mlu-ops-dev/mlu-ops/bangc-ops/test/mlu_op_gtest/pb_gtest/src/gtest/mlu_op_gtest.cpp:125: Failure
```

Test Point-2: `MLU500 box tensor should less than 2G nums`. Acceptance Standard: `Normal error`.
```bash
[2023-10-18 14:34:4] [MLUOP] [Error]:[mluOpBoxIouRotated] overflow max supported tensor num 2147483647, now tensor's total num is 2147483650. 
[2023-10-18 14:34:4] [MLUOP] [Error]:"MLUOP_STATUS_NOT_SUPPORTED in mluOpBoxIouRotated(handle_, mode, aligned, box1, dev_box1, box2, dev_box2, ious, dev_ious)"
[2023-10-18 14:34:4] [MLUOP] [Warning]:Executor: reset cnrt and go on running, this may caused by cnrt failed.
/projs/platform/liuduanhui/mluops-dev/mlu-ops/bangc-ops/test/mlu_op_gtest/pb_gtest/src/gtest/mlu_op_gtest.cpp:125: Failure
```

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU590

|         shape                | baseline(us) |  current(us)  | percent of <br>time reduced |
|------------------------------|----------|--------|--------|
|[32,5] aligned=1          | 82       | 24     | 70.73% |
|[640,5]  aligned=1         | 227.9    | 35     | 84.64% |
|[2400,5]  aligned=1        | 1082.05  | 34.44  | 96.82% |
|[18430,5] aligned=1       | 22225.76 | 440.75 | 98.02% |
|[38000,5] aligned=1       | 12217.35 | 246.38 | 97.98% |
|[160,5]  aligned=1         | 165      | 34     | 79.39% |
|[260,5]  aligned=1         | 108      | 29     | 73.15% |
|[570,5] aligned=1         | 190.31   | 31     | 83.71% |
|[2048,5] aligned=1        | 176.02   | 26     | 85.23% |
|[18432,5] aligned=1       | 599.04   | 68     | 88.65% |
|[250,5],[5,5] aligned=0   | 2014.03  | 1877.01| 6.80%  |
|[60,5],[30,5]  aligned=0   | 1909.06  | 697.82 | 63.45% |
|[10,5],[30,5]  aligned=0   | 328      | 208    | 36.59% |
| [5,5],[320,5] aligned=0  | 828      | 46     | 94.44% |

### 3.4 Summary Analysis

The performance improve by 6.8 % to 98% on 500.

### CODE REVIEW

File: bangc-ops/kernels/box_iou_rotated/box_iou_rotated_utils.h [hunk 1]
```diff
@@ -479,13 +496,121 @@ __mlu_func__ void convexHullGraham(
...
   __bang_minpool((T *)temp3_ram, (T *)temp_long_1, actual_compute_box_num, 1,
                  24, 1, 24, 1, 24);
-
   __bang_mul((T *)temp3_ram, (T *)temp3_ram, (T *)valid_box,
              actual_compute_box_num);
 
+#if __BANG_ARCH__ >= 592
+  // Step2. All points subtract starting-point (for sorting in the next step)
+  __bang_cycle_sub((T *)intersect_pts_x, (T *)intersect_pts_x, (T *)temp3_ram,
+                   24 * actual_compute_box_num, actual_compute_box_num);
```
Comment (PetrelYy) [inline bangc-ops/kernels/box_iou_rotated/box_iou_rotated_utils.h:509]: 代码里多处用到24 1 3， 建议用变量替换魔鬼数字

File: docs/bangc-docs/design_docs/box_iou_rotated/box_iou_rotated.md [hunk 2]
```diff
@@ -145,7 +144,15 @@ mluOpStatus_t MLUOP_WIN_API mluOpBoxIouRotated(mluOpHandle_t handle,
 3. 计算`new_box_pts`，向量化地计算新的`x_ctr`, `y_ctr`，但是 width/height/theta 都可以来源于输入，无需复制
 4. 通过`new_pts`的数据，计算旋转后的顶点`rotated_vertices`，通过旋转后的顶点，计算每条边的`vector`的表示，通过 3 个过程，计算 24 种可能的交点。由于向量化计算，每组 box-pair 的有效交点个数和类型都不同，需要设置`valid_pts`对应位置是否为 true，标记该位置是否为有效交点。
 5. 通过上述过程得到的有效交点，如果`nums_in`对应大于 2，代表有 2 个以上的交点，可以计算交叠面积，否则设置对应位置的`valid_box`为 false，设置 ious=0.
-6. 用 Convex-hull-graham 顶点扫描法，标量计算每一组 24 个交点，进行排序、筛选得出凸包形状的顶点集合，最后使用`polygon_area`函数计算有效 box 的 iou 面积。然后再根据`mode`的不同，做不同分母的除法（区分 arch）。
+6. 用 Convex-hull-graham 顶点扫描法，标量计算每一组 24 个交点，进行排序、筛选得出凸包形状的顶点集合。（区分 arch）
+ - 1. 300 系列顶点间的顺序有cross_product值比较来进行排序。
+ - 2. 500 系列
+  - 1. 两个box相交得出的多边形只能是凸包，排序正确情况下，没有凹点需要筛选，可以省略筛选点的步骤。
```
Comment (PetrelYy) [inline docs/bangc-docs/design_docs/box_iou_rotated/box_iou_rotated.md:150]: 这里markdown显示有些问题，既有 * ，又有序号

File: docs/bangc-docs/design_docs/box_iou_rotated/box_iou_rotated.md [hunk 3]
```diff
@@ -145,7 +144,15 @@ mluOpStatus_t MLUOP_WIN_API mluOpBoxIouRotated(mluOpHandle_t handle,
...
+  - 2. 按照各点与x正半轴形成的角度，逆时针排序。    
+      ![box_iou_rotated](./res/sort_vertices1.png)
+  - 3. 设点p0(x, y)， 计算 f(p0) = abs(x) * x / (x * x + y * y), cos函数在[0, π]区间（第一二象限）单调递减，如下图所示。通过比较两个顶点 `cosθ`平方的大小（保留正负号），来判断两个点的大小关系。
+      ![box_iou_rotated](./res/cosine1.png)
```
Comment (PetrelYy) [inline docs/bangc-docs/design_docs/box_iou_rotated/box_iou_rotated.md:154]: 显示图片没换行

File: bangc-ops/kernels/box_iou_rotated/box_iou_rotated_utils.h (hunk 1 above)
Comment (DanieeelLiu) [reply bangc-ops/kernels/box_iou_rotated/box_iou_rotated_utils.h:509]: 24 有上下文，与前面风格相同，3 已修改

File: docs/bangc-docs/design_docs/box_iou_rotated/box_iou_rotated.md (hunk 2 above)
Comment (DanieeelLiu) [reply docs/bangc-docs/design_docs/box_iou_rotated/box_iou_rotated.md]: Fixed

File: docs/bangc-docs/design_docs/box_iou_rotated/box_iou_rotated.md (hunk 3 above)
Comment (DanieeelLiu) [reply docs/bangc-docs/design_docs/box_iou_rotated/box_iou_rotated.md]: Fixed

**Review by mahxn0 — APPROVED** [review]

**Review by duzekunKTH — APPROVED** [review]

File: bangc-ops/kernels/box_iou_rotated/box_iou_rotated_utils.h [hunk 4]
```diff
@@ -479,13 +496,118 @@ __mlu_func__ void convexHullGraham(
...
+
+  // Step3. Sorting according to angles. The angle become larger from 0 to 180
+  // degree in anti-clockwise direction. The cos value of the value change from
+  // 1 to -1. Cos(theta) = fabs(x)*x / x*x + y*y
+  dot2d<T>((T *)dist_ram, (T *)intersect_pts_x, (T *)intersect_pts_y,
+           (T *)intersect_pts_x, (T *)intersect_pts_y,
+           24 * actual_compute_box_num, (T *)temp_long_3);
+
+  // make sure points don't overlap with each other
+  __bang_mul((T *)temp_long_1, (T *)dist_ram, (T *)valid_pts, total_points);
```
Comment (Unireverse) [inline bangc-ops/kernels/box_iou_rotated/box_iou_rotated_utils.h:525]: 虽然但是，这个命名没看懂想干嘛hhh，可以不改

**Review by Unireverse — APPROVED** [review]

File: bangc-ops/kernels/box_iou_rotated/box_iou_rotated.cpp [hunk 5]
```diff
@@ -149,10 +149,10 @@ mluOpBoxIouRotated(mluOpHandle_t handle, const int mode, const bool aligned,
...
       return MLUOP_STATUS_BAD_PARAM;
     }
-    if (box1_desc->dims[0] > MAX_BOX_NUM) {
+    if (handle->arch < 592 && box1_desc->dims[0] > MAX_BOX_NUM) {
       LOG(ERROR) << "[mluOpBoxIouRotated] Check failed: If it is aligned mode, "
-                 << "box1_desc->dims[0] should less than or equal to "
-                 << "10,000,000. But now is " << box1_desc->dims[0] << ".";
+                 << "on MLU300 box1_desc->dims[0] should less than or equal to "
+                 << "10,000,000 . But now is " << box1_desc->dims[0] << ".";
       return MLUOP_STATUS_BAD_PARAM;
```
[the remark this answers is not in the document]
Comment (DanieeelLiu) [reply bangc-ops/kernels/box_iou_rotated/box_iou_rotated.cpp]: Fixed

**Pull request merged**